### PR TITLE
[entropy_src/rtl] Move the 1-4 esbit packer

### PR DESCRIPTION
**Pull request by h-filali**
This PR moves the esbit packer FIFO in front of the postht FIFO.
For a more detailed explanation have a look at the commit messages.

Closes [#20954](https://github.com/lowRISC/opentitan/issues/20954)

### CODE REVIEW

File: hw/ip/entropy_src/rtl/entropy_src_core.sv [hunk 1]
```diff
@@ -1033,8 +1033,8 @@ module entropy_src_core import entropy_src_pkg::*; #(
 
   assign sfifo_esrng_clr   = ~es_delayed_enable;
   assign sfifo_esrng_wdata = es_rng_bus;
-  assign sfifo_esrng_pop   = sfifo_esrng_not_empty & (rng_bit_en ? pfifo_esbit_not_full :
-                                                                   pfifo_postht_not_full );
+  assign sfifo_esrng_pop   = rng_bit_en ? sfifo_esrng_not_empty && pfifo_esbit_not_full :
+                             sfifo_esrng_not_empty && pfifo_postht_not_full;
```
Comment (vogelpi) [inline hw/ip/entropy_src/rtl/entropy_src_core.sv:1037]: I think this change is not really needed (functionality doesn't change) but it makes the code harder to read. Would you mind changing this back please?

File: hw/ip/entropy_src/rtl/entropy_src_core.sv [hunk 2]
```diff
@@ -1085,7 +1085,7 @@ module entropy_src_core import entropy_src_pkg::*; #(
     .depth_o    ()
   );
 
-  assign pfifo_esbit_push = rng_bit_en && sfifo_esrng_not_empty;
+  assign pfifo_esbit_push = rng_bit_en && sfifo_esrng_not_empty && pfifo_esbit_not_full;
```
Comment (vogelpi) [inline hw/ip/entropy_src/rtl/entropy_src_core.sv:1088]: What is the motivation of this change? Do we really need this?

My problem with this is the following: right now the code just writes the FIFO even if it's full. Of course this is bad but it's captured with an err_code that software can inspect. I.e., if this happens, software will see this and can restart the entropy complex.

If we don't push at this point, the alert will not fire. In this particular case, the esrng FIFO will become full and eventually disable the noise source. The noise source is restarted and everything continues without software getting notified and without repeating the startup test which is very bad from a validation viewpoint. Of course we need to change this, but at this point I would not do such changes unless it's really required.

File: hw/ip/entropy_src/rtl/entropy_src_core.sv [hunk 3]
```diff
@@ -1097,8 +1097,9 @@ module entropy_src_core import entropy_src_pkg::*; #(
 
   // select source for health testing
 
-  assign health_test_esbus     = pfifo_postht_wdata;
-  assign health_test_esbus_vld = pfifo_postht_push & pfifo_postht_not_full & ~pfifo_postht_clr;
+  assign health_test_esbus     = sfifo_esrng_rdata;
+  assign health_test_esbus_vld = rng_bit_en ? pfifo_esbit_push && !pfifo_esbit_clr :
+                                 pfifo_postht_push && !pfifo_postht_clr;
```
Comment (vogelpi) [inline hw/ip/entropy_src/rtl/entropy_src_core.sv:1102]: please indent the lower line for better readability:
```sv
 assign health_test_esbus_vld = rng_bit_en ? pfifo_esbit_push && !pfifo_esbit_clr :
                                             pfifo_postht_push && !pfifo_postht_clr;
```

File: hw/ip/entropy_src/rtl/entropy_src_core.sv [hunk 4]
```diff
@@ -1097,8 +1097,9 @@ module entropy_src_core import entropy_src_pkg::*; #(
 
   // select source for health testing
```
Comment (vogelpi) [inline hw/ip/entropy_src/rtl/entropy_src_core.sv:1098]: This comment is now outdated. Please adapt it accordingly.

File: hw/ip/entropy_src/rtl/entropy_src_core.sv [hunk 5]
```diff
@@ -1182,8 +1183,12 @@ module entropy_src_core import entropy_src_pkg::*; #(
   assign hw2reg.extht_lo_thresholds.bypass_thresh.d = extht_lo_bypass_threshold_oneway;
 
 
-
-  assign health_test_window = es_bypass_mode ? health_test_bypass_window : health_test_fips_window;
+  // Multiply the respective windows by four if we are using the single lane mode.
+  // In single lane mode 4 times as many symbols are tested for the same amount of entropy.
+  assign health_test_window = (es_bypass_mode && rng_bit_en) ? health_test_bypass_window << 2 :
```
Comment (vogelpi) [inline hw/ip/entropy_src/rtl/entropy_src_core.sv:1188]: You'll also need to update the width of the counter primitive (u_prim_count_window_cntr). Currently it's 16-bit wide but it needs to be 18 bits I guess.

File: hw/ip/entropy_src/rtl/entropy_src_core.sv [hunk 6]
```diff
@@ -2358,8 +2371,8 @@ module entropy_src_core import entropy_src_pkg::*; #(
     .depth_o    ()
   );
 
-  assign pfifo_postht_push = rng_bit_en ? pfifo_esbit_not_empty :
-                             sfifo_esrng_not_empty;
+  assign pfifo_postht_push = rng_bit_en ? pfifo_esbit_not_empty && pfifo_postht_not_full :
+                             sfifo_esrng_not_empty && pfifo_postht_not_full;
```
Comment (vogelpi) [inline hw/ip/entropy_src/rtl/entropy_src_core.sv:2375]: Same here: it's probably better not to include the full signal into the decision. It would be better from design perspective but at least I don't know all the implications this might have on the design.

File: hw/ip/entropy_src/rtl/entropy_src_repcnt_ht.sv [hunk 7]
```diff
@@ -102,15 +106,26 @@ module entropy_src_repcnt_ht #(
...
+                            (rng_bit_sel_i == 2'h2) ? rep_cnt_fail[2] :
+                            rep_cnt_fail[3];
+
   // For the purposes of failure pulse generation, we want to sample
   // the test output for only one cycle and do it immediately after
   // the counter has been updated.
   assign fail_sample_mask_d = entropy_bit_vld_i;
-  assign fail_sampled       = |rep_cnt_fail & fail_sample_mask_q;
+  assign fail_sampled       = rng_bit_en_i ? rng_bit_cnt_fail & fail_sample_mask_q :
+                              |rep_cnt_fail & fail_sample_mask_q;
```
Comment (vogelpi) [inline hw/ip/entropy_src/rtl/entropy_src_repcnt_ht.sv:124]: Please indent the lower line for better readability:
```sv
  assign fail_sampled       = rng_bit_en_i ? rng_bit_cnt_fail & fail_sample_mask_q :
                                                |rep_cnt_fail & fail_sample_mask_q;
```
Or better move common term out of the ternary expression:
```sv
  assign fail_sampled = fail_sample_mask_q &
      (rng_bit_en_i ? rng_bit_cnt_fail : rep_cnt_fail);
```

**Review by vogelpi — COMMENTED** [review]
Thanks for working on this @h-filali , this looks mostly good!

We'll also need to take care of the documentation including the block diagram. Since the RTL code follows the block diagram, the code section related to the 1-to-4 esbit packer FIFO should maybe be moved down a bit. As it will however complicate the review a lot, I would say let's first iron out the other stuff.

File: hw/ip/entropy_src/rtl/entropy_src_repcnt_ht.sv (hunk 7 above)
Comment (h-filali) [reply hw/ip/entropy_src/rtl/entropy_src_repcnt_ht.sv:124]: I'll change it, thanks!

File: hw/ip/entropy_src/rtl/entropy_src_core.sv (hunk 2 above)
Comment (h-filali) [reply hw/ip/entropy_src/rtl/entropy_src_core.sv:1088]: You're absolutely right. I did this because it just made sense to add this, as it would in the usual case, because why would we push into a full FIFO.
However, I forgot to think about the errors we have for this. So you are absolutely right, the added signal should not be there and should not be changed.

File: hw/ip/entropy_src/rtl/entropy_src_core.sv (hunk 5 above)
Comment (h-filali) [reply hw/ip/entropy_src/rtl/entropy_src_core.sv:1188]: Nice catch, thanks!

**Comment by h-filali** [comment]
Thanks for reviewing @vogelpi, there were some really important insights in there. I have addressed all your comments. Let's see what the CI has to say and move on from there :nerd_face:

File: hw/ip/entropy_src/dv/env/entropy_src_scoreboard.sv [hunk 8]
```diff
@@ -1069,8 +1095,10 @@ class entropy_src_scoreboard extends cip_base_scoreboard#(
     seed_tl_read_cnt = 0;
 
     for (int i = 0; i < RNG_BUS_WIDTH; i++) begin
-      `uvm_info(`gfn, "Clearing REPCNTS cntr", UVM_DEBUG)
-      repcnt[i]       = 1;
+      `uvm_info(`gfn, "Set REPCNTS cntr", UVM_DEBUG)
+      repcnt[i] = (rst_type == HardReset) ? 0 :
+                  (rst_type == Enable) ? 1 :
+                  repcnt[i];
```
Comment (vogelpi) [inline hw/ip/entropy_src/dv/env/entropy_src_scoreboard.sv:1101]: Is this change related to this PR?

File: hw/ip/entropy_src/doc/programmers_guide.md [hunk 9]
```diff
@@ -92,6 +92,24 @@ Note that once the field `ES_ROUTE` is set to `kMultiBitBool4True`, no entropy i
...
+## Single RNG Bit Mode
+
+The PTRNG entropy source delivers a stream of symbols to ENTROPY_SRC.
+Each of these symbols contains multiple bits.
+Firmware can determine whether ENTROPY_SRC should utilize all or only one of the bits that make up a symbol.
+Per default, the `RNG_BIT_ENABLE` field in the [`CONF`](registers.md#conf) register is set to `kMultiBitBool4False` and the whole symbol is used.
+To enable the single RNG bit mode `RNG_BIT_ENABLE` has to be set to `kMultiBitBool4True`.
+To select which specific bit should be used, the `RNG_BIT_SEL` field in the [`CONF`](registers.md#conf) register must be set to the number of the desired lane.
+When ENTROPY_SRC is in single RNG bit mode, only a subset of the health tests continue to be applicable.
+For this reason the thresholds, for both the bypass and the FIPS values, must be set accordingly.
```
Comment (vogelpi) [inline hw/ip/entropy_src/doc/programmers_guide.md:104]: Nit: For this reason, the thresholds for both the bypass and the FIPS values must be set accordingly.

File: hw/ip/entropy_src/doc/programmers_guide.md [hunk 10]
```diff
@@ -92,6 +92,24 @@ Note that once the field `ES_ROUTE` is set to `kMultiBitBool4True`, no entropy i
...
+The repetition count test, markov test and adaptive proportion test can all still be performed on a single lane.
```
Comment (vogelpi) [inline hw/ip/entropy_src/doc/programmers_guide.md:105]: Nit: markov -> Markov

File: hw/ip/entropy_src/doc/programmers_guide.md [hunk 11]
```diff
@@ -92,6 +92,24 @@ Note that once the field `ES_ROUTE` is set to `kMultiBitBool4True`, no entropy i
...
+However, to get the same amount of entropy we now must collect four times as many individual samples from PTRNG.
```
Comment (vogelpi) [inline hw/ip/entropy_src/doc/programmers_guide.md:106]: However, to get the same number of entropy bits, we now must collect four times as many individual symbols from the PTRNG.

File: hw/ip/entropy_src/doc/programmers_guide.md [hunk 12]
```diff
@@ -92,6 +92,24 @@ Note that once the field `ES_ROUTE` is set to `kMultiBitBool4True`, no entropy i
...
 
+## Single RNG Bit Mode
+
+The PTRNG entropy source delivers a stream of symbols to ENTROPY_SRC.
+Each of these symbols contains multiple bits.
+Firmware can determine whether ENTROPY_SRC should utilize all or only one of the bits that make up a symbol.
+Per default, the `RNG_BIT_ENABLE` field in the [`CONF`](registers.md#conf) register is set to `kMultiBitBool4False` and the whole symbol is used.
+To enable the single RNG bit mode `RNG_BIT_ENABLE` has to be set to `kMultiBitBool4True`.
+To select which specific bit should be used, the `RNG_BIT_SEL` field in the [`CONF`](registers.md#conf) register must be set to the number of the desired lane.
+When ENTROPY_SRC is in single RNG bit mode, only a subset of the health tests continue to be applicable.
```
Comment (vogelpi) [inline hw/ip/entropy_src/doc/programmers_guide.md:103]: Nit: add a line break here. Start a new paragraph for discussing the health tests.

File: hw/ip/entropy_src/doc/programmers_guide.md [hunk 13]
```diff
@@ -92,6 +92,24 @@ Note that once the field `ES_ROUTE` is set to `kMultiBitBool4True`, no entropy i
...
+For this reason the thresholds, for both the bypass and the FIPS values, must be set accordingly.
+The repetition count test, markov test and adaptive proportion test can all still be performed on a single lane.
+However, to get the same amount of entropy we now must collect four times as many individual samples from PTRNG.
+Therefore, the threshold values for the above tests need to be set as desired.
+The other tests, namely, the symbol repetition count test and the bucket test are not applicable to a single lane.
```
Comment (vogelpi) [inline hw/ip/entropy_src/doc/programmers_guide.md:108]: The symbol repetition count test and the bucket test are not applicable to a single lane.
they need to be disabled by setting the thresholds to the maximum value.

File: hw/ip/entropy_src/rtl/entropy_src_core.sv (hunk 2 above)
Comment (vogelpi) [reply hw/ip/entropy_src/rtl/entropy_src_core.sv]: Thanks for reverting this!

**Review by vogelpi — APPROVED** [review]
This looks great!

I have some nits regarding the documentation. Thanks @h-filali for doing this.

File: hw/ip/entropy_src/rtl/entropy_src_core.sv [hunk 14]
```diff
@@ -59,6 +59,7 @@ module entropy_src_core import entropy_src_pkg::*; #(
   localparam int PostHTWidth = 32;
   localparam int RngBusWidth = 4;
   localparam int HalfRegWidth = 16;
+  localparam int WindowCntrWidth = 18;
```
[the remark this answers is not in the document]
Comment (h-filali) [reply hw/ip/entropy_src/rtl/entropy_src_core.sv:62]: Agreed, I'll change it

File: hw/ip/entropy_src/dv/env/entropy_src_scoreboard.sv (hunk 8 above)
Comment (vogelpi) [reply hw/ip/entropy_src/dv/env/entropy_src_scoreboard.sv:1101]: Okay, thanks for explaining :-)
